vfs: nil pointer dereference in memFile.Write

Seen on CI https://travis-ci.org/cockroachdb/pebble/jobs/656363897?utm_medium=notification&utm_source=email

I haven't been able to reproduce running `make stressmeta`. CI doesn't provides access to the OPTIONS or ops used by the metamorphic test. Probably should fix that.

```
--- FAIL: TestMeta (14.20s)
    --- FAIL: TestMeta/random-010 (0.96s)
        meta_test.go:206: exit status 2
            _meta/200228-173559.766/random-010/history
            
            panic: runtime error: invalid memory address or nil pointer dereference
            [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x53e44b]
            
            goroutine 4209 [running]:
            github.com/cockroachdb/pebble/vfs.(*memFile).Write(0xc0002da9c0, 0xc000648e38, 0x0, 0x7238, 0x0, 0x0, 0x0)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/vfs/mem_fs.go:639 +0x5b
            github.com/cockroachdb/pebble/internal/record.(*Writer).writePending(0xc000648000)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/record.go:494 +0x98
            github.com/cockroachdb/pebble/internal/record.(*Writer).Close(0xc000648000, 0x8d63e0, 0xc000e41c70)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/record.go:501 +0x33
            github.com/cockroachdb/pebble.(*versionSet).createManifest(0xc000784180, 0xc0000a61b0, 0x27, 0xfa5, 0xf9f, 0xfa6, 0x0, 0x0)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/version_set.go:515 +0x574
            github.com/cockroachdb/pebble.(*versionSet).logAndApply.func1(0xc000784180, 0xc000a4c3c0, 0xc000e87110, 0xc0000cf9c0, 0xc0000cf9a0, 0xfa5, 0xf9f, 0xfa6, 0x42f, 0x8dec00, ...)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/version_set.go:365 +0x8b7
            github.com/cockroachdb/pebble.(*versionSet).logAndApply(0xc000784180, 0x42f, 0xc000a4c3c0, 0xc00075e150, 0x8dec00, 0xc0002da240, 0xc0000cfc18, 0x0, 0x0)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/version_set.go:413 +0x1ad
            github.com/cockroachdb/pebble.(*DB).compact1(0xc000784000, 0xc00000c3c0, 0x0, 0x0, 0x0)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/compaction.go:1119 +0x8cd
            github.com/cockroachdb/pebble.(*DB).compact.func1(0x8dd4c0, 0xc00075e120)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/compaction.go:1065 +0xbe
            runtime/pprof.Do(0x8dd4c0, 0xc00075e120, 0xc0000b6120, 0x1, 0x1, 0xc000593798)
            	/home/travis/.gimme/versions/go1.13.8.linux.amd64/src/runtime/pprof/runtime.go:37 +0xf8
            github.com/cockroachdb/pebble.(*DB).compact(0xc000784000, 0xc00000c3c0, 0x0)
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/compaction.go:1062 +0xa2
            created by github.com/cockroachdb/pebble.(*DB).maybeScheduleCompaction
            	/home/travis/gopath/src/github.com/cockroachdb/pebble/compaction.go:1056 +0x3c2
```

--- a/compaction.go
+++ b/compaction.go
@@ -1050,6 +1050,12 @@ func (d *DB) maybeScheduleCompaction() {
 	d.mu.versions.logLock()
 	defer d.mu.versions.logUnlock()
 
+	// Check for the closed flag again, in case the DB was closed while we were
+	// waiting for logLock().
+	if atomic.LoadInt32(&d.closed) != 0 {
+		return
+	}
+
 	env := compactionEnv{
 		bytesCompacted:          &d.bytesCompacted,
 		earliestUnflushedSeqNum: d.getEarliestUnflushedSeqNumLocked(),

--- a/db_test.go
+++ b/db_test.go
@@ -17,6 +17,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/pebble/internal/base"
+	"github.com/cockroachdb/pebble/sstable"
 	"github.com/cockroachdb/pebble/vfs"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/exp/rand"
@@ -1033,6 +1034,34 @@ func TestDBConcurrentCommitCompactFlush(t *testing.T) {
 
 	if err := d.Close(); err != nil {
 		t.Fatal(err)
+	}
+}
+
+func TestDBConcurrentCompactClose(t *testing.T) {
+	// Test closing while a compaction is ongoing. This ensures compaction code
+	// detects the close and finishes cleanly.
+	mem := vfs.NewMem()
+	for i := 0; i < 100; i++ {
+		d, err := Open("", &Options{
+			FS:                       mem,
+			MaxConcurrentCompactions: 2,
+		})
+		require.NoError(t, err)
+
+		// Ingest a series of files containing a single key each. As the outer
+		// loop progresses, these ingestions will build up compaction debt
+		// causing compactions to be running concurrently with the close below.
+		for j := 0; j < 10; j++ {
+			path := fmt.Sprintf("ext%d", j)
+			f, err := mem.Create(path)
+			require.NoError(t, err)
+			w := sstable.NewWriter(f, sstable.WriterOptions{})
+			require.NoError(t, w.Set([]byte(fmt.Sprint(j)), nil))
+			require.NoError(t, w.Close())
+			require.NoError(t, d.Ingest([]string{path}))
+		}
+
+		require.NoError(t, d.Close())
 	}
 }
 

--- a/version_set.go
+++ b/version_set.go
@@ -270,7 +270,8 @@ func (vs *versionSet) close() error {
 // logLock locks the manifest for writing. The lock must be released by either
 // a call to logUnlock or logAndApply.
 //
-// DB.mu must be held when calling this method.
+// DB.mu must be held when calling this method, but the mutex may be dropped and
+// re-acquired during the course of this method.
 func (vs *versionSet) logLock() {
 	// Wait for any existing writing to the manifest to complete, then mark the
 	// manifest as busy.

